### DISCUSSION TITLE
ArcfaceLoss vs ArcfaceLossSimple

### DISCUSSION THREAD

**Discussion by nadongjin** [leondgarse/Keras_insightface#85] [Q&A]
Hello
I have another question while looking at your code.
The **ArcfaceLoss** and **ArcfaceLossSimple** of your code are actually the same as the current learning criteria, right?
Because ArcfaceLossSimple's cos(y_pred_values+margin) = cos(y_pred_values) * cos(margin) - sin(y_pred_vals) * sin(margin), ArcfaceLoss's  theta = tf.cos(tf.acos(y_pred_vals) + self.margin2) 

margin==margin2==0.5

class ArcfaceLoss(tf.keras.losses.Loss):
    <SKIP>
    def call(self, y_true, norm_logits):
        pick_cond = tf.where(y_true != 0)
        y_pred_vals = tf.gather_nd(norm_logits, pick_cond)
        if self.margin1 == 1.0 and self.margin2 == 0.0 and self.margin3 == 0.0:
            theta = y_pred_vals
        **elif self.margin1 == 1.0 and self.margin3 == 0.0:
            theta = tf.cos(tf.acos(y_pred_vals) + self.margin2)**
        else:
            theta = tf.cos(tf.acos(y_pred_vals) * self.margin1 + self.margin2) - self.margin3

class ArcfaceLossSimple(tf.keras.losses.Loss):
  <SKIP>
    def call(self, y_true, norm_logits):
        pick_cond = tf.where(y_true != 0)
        y_pred_vals = tf.gather_nd(norm_logits, pick_cond)
        **theta = y_pred_vals * self.margin_cos - tf.sqrt(1 - tf.pow(y_pred_vals, 2)) * self.margin_sin**
        theta_valid = tf.where(y_pred_vals > self.threshold, theta, self.theta_min - theta)
        arcface_logits = tf.tensor_scatter_nd_update(norm_logits, pick_cond, theta_valid) * self.scale
        return tf.keras.losses.categorical_crossentropy(y_true, arcface_logits, from_logits=self.from_logits, label_smoothing=self.label_smoothing)

**Comment by leondgarse** [comment]
Yes, `ArcfaceLossSimple` is just a simple version of `ArcfaceLoss`, that using only `margin2`. Their `theta` results are same.

**Reply by nadongjin** [reply]
Thank you for your answer.

Also, this is really my personal opinion.
Wouldn't it be better if the norm value of 512 embedding was included in the similarity when measuring similarity with the learned magface?
The problem is that I don't know how to add or subtract the norm value. :)

The reason for this is that in the case of sensetime, it seems to include something other than embedding when measuring similarity.

thanks.

**Reply by nadongjin** [reply]
Uh, there's something similar in the Magface paper.
![image](https://user-images.githubusercontent.com/14832440/155683767-42692455-1af2-4f50-8cb9-bf6fd457b2cd.png)

**Reply by leondgarse** [reply]
I think that's currently the default way evaluating `IJB`s:
![Selection_447](https://user-images.githubusercontent.com/5744524/155687768-e5bff02f-89b9-4274-bd83-ff150e710dd0.png)
That default `use_norm_score=False` will not normalize features in the `process_embeddings` step, and then the raw feature values are summed and normalized for further validation. Looks like the suggested `f+` method.
In my understanding, this norm value can be used in production environment filtering out low quality face images. But using it for similarity testing... maybe your previous separating `sum_cosine_similarity` can apply it, by setting different weights for top and bottom half:
  ```py
  def sum_cosine_similaty(vector1, vector2):
      emb1 = vector1 / np.linalg.norm(vector1)
      emb2 = vector2 / np.linalg.norm(vector2)

      vector1_top_norm = np.linalg.norm(vector1[:256])
      vector2_top_norm = np.linalg.norm(vector2[:256])
      emb3 = vector1[:256] / vector1_top_norm
      emb4 = vector2[:256] / vector2_top_norm

      vector1_bottom_norm = np.linalg.norm(vector1[256:])
      vector2_bottom_norm = np.linalg.norm(vector2[256:])
      emb5 = vector1[256:] / vector1_bottom_norm
      emb6 = vector2[256:] / vector2_bottom_norm

      top_weight = vector1_top_norm * vector2_top_norm
      bottom_weight = vector1_bottom_norm * vector2_bottom_norm

      # sum1 = np.sum(emb1 * emb2) #all cosine similarity
      # sum2 = np.sum(emb3 * emb4) # front :256 cosine similarity
      # sum3 = np.sum(emb5 * emb6) # back 256: cosine similarity
      # return sum1 + sum2 + sum3
      sum2 = np.sum(emb3 * emb4) * top_weight # front :256 cosine similarity
      sum3 = np.sum(emb5 * emb6) * bottom_weight # back 256: cosine similarity
      return (sum2 + sum3) / (top_weight + bottom_weight)

  aa, bb = tf.random.uniform([512]), tf.random.uniform([512])
  sum_cosine_similaty(aa, bb)
  # 0.7597335
  sum_cosine_similaty(aa, aa)
  # 0.9999999
  ```
  **Simply**:
  ```py
  def sum_cosine_similarity(vector1, vector2):
      # emb1 = vector1 / np.linalg.norm(vector1)
      # emb2 = vector2 / np.linalg.norm(vector2)

      top_weight = np.linalg.norm(vector1[:256]) * np.linalg.norm(vector2[:256])
      bottom_weight = np.linalg.norm(vector1[256:]) * np.linalg.norm(vector2[256:])
      sum2 = np.sum(vector1[:256] * vector2[:256]) # front :256 cosine similarity
      sum3 = np.sum(vector1[256:] * vector2[256:]) # back 256: cosine similarity
      return (sum2 + sum3) / (top_weight + bottom_weight)
  ```

**Reply by nadongjin** [reply]
Thank you for your great answer.
Have you ever used SCFRD-34GF?
I tried to change the insightface's pytorch to keras, onnx, but failed.
![image](https://user-images.githubusercontent.com/14832440/155752697-a0a22e13-50a3-4225-9dc6-1273ee65fc90.png)

**Reply by leondgarse** [reply]
Not yet. They are providing [scrfd2onnx.py](https://github.com/deepinsight/insightface/blob/master/detection/scrfd/tools/scrfd2onnx.py), is that not working?

**Reply by nadongjin** [reply]
Your code works well without any problems.
However, I think face detection and landmark are more important than anything else in actual accuracy that does not provide a separate landmark such as IJBB and IJBC, so I would like to try this.
My ms1mv3 dataset organization work is 75% completed. Please wait a little longer.
I think your learning method is SOTA accuracy with only limited datasets.
Since the category_cross_entropy itself generally uses the loss average, I think there will be not a big problem even in the wrong datasets.
However, if you learn with a good datasets, unnecessary loss will be reduced, so I think it will bring the inside closer and the outside farther away.
In fact, about 2.5% of the ms1mv3 datasets were duplicated or had problems.
The conclusion is that if you are given great data, your model will not be able to keep up with anyone.
Already, Chinese, Korean, or global facial recognition companies are enormous and have excellent data. In fact, some companies can buy data with money.
But you and I can't do that.
For example, your model is generally excellent, but I think it is the cause of a slight decline in performance compared to global companies in terms of accuracy of Asians and blacks.

**Reply by leondgarse** [reply]
Ya, you probably right.
- By `accuracy of Asians and blacks`, are you referring `MR-ALL` validation in `insightface` model zoo? I haven't run that test myself.
- Using a cleaner dataset may be able to apply a heavier regularizer, like larger weight_decay, larger randaug, or larger margin.
- The dataset glitch may explain why `Triplet` performing bad when fine-tuning an Arcface trained model.

**Reply by nadongjin** [reply]
Hello
I have an idea.
When I did the accuracy test, I confirmed that the accuracy was improved by flipping on one image and then averaging it.(Of course, this is included in your code.)
In common, models with nist's good accuracy tend to take up a lot of memory when making inference.
But the model size is not big.
Therefore, I think it would be good to improve performance by applying another random_status differently like the image flip applied during learning. : )
thanks..

**Reply by nadongjin** [reply]
[example]
![image](https://user-images.githubusercontent.com/14832440/156015324-03abdd0d-6711-4bd4-b072-274579b4e525.png)

**Reply by leondgarse** [reply]
Do you mean a random process averaging the augmented images during training?
- For augment like `random_brightness / random_contrast`, it may not help. Take `random_brightness` alone for instance, averaging it with original image may just result an image more like original image, works like a smaller brightness `delta`.
- For averaging flipped image with original image, I added some commented code in [flip average test](https://github.com/leondgarse/Keras_insightface/commit/2289b10cd0dddc7ac651b03ac48d356a00d6fdad). If use it instead of `img = tf.image.random_flip_left_right(img)`, will set a `1/3` probability for either original / flipped / averaged image. I think it may help, but still not sure, as some averaged results went weird.
  ![Selection_448](https://user-images.githubusercontent.com/5744524/156281568-d5f421bf-80f3-48fd-bfe5-3b02c511c0fa.png)
- For averaging `random_crop`, can be applied similarly like `random_flip`.
